server/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableAwareStreamInput.java

```diff
@@ -28,6 +32,21 @@ public NamedWriteableAwareStreamInput(StreamInput delegate, NamedWriteableRegist
         return readNamedWriteable(categoryClass, name);
     }
 
+    @Override
+    public <T extends NamedWriteable> List<T> readNamedWriteableList(Class<T> categoryClass) throws IOException {
+        int count = readArraySize();
+        if (count == 0) {
+            return Collections.emptyList();
+        }
+        final Map<String, Writeable.Reader<?>> readers = namedWriteableRegistry.getReaders(categoryClass);
+        List<T> builder = new ArrayList<>(count);
+        for (int i = 0; i < count; i++) {
+            final String name = readString();
+            builder.add(NamedWriteableRegistry.getReader(categoryClass, name, readers).read(this));
```

Comment:
Would it save anything to check whether `name` matches the `name` from the previous iteration and re-use the previous reader instead of doing another lookup? I expect in practice we see the same `name` repeatedly.

Reply:
>  I expect in practice we see the same name repeatedly.

I'm not so sure about that. In queries for example (which is pretty the one thing where this gets really hot) that doesn't seem to be the case but could be I'm wrong there. The thing is, you trade a string comparison for a map lookup. That saves you hashing `name` (which is a freshly read string so you have to compute the hash code) if you're right, but if you're wrong you get hashing and 2 comparisons. Given how cheap the map lookup is for the immutable maps you'd have to be right a ton here to justify this optimization.
Also, by adding that branching you made the loop more complicated which might have some negative side effects as well in terms of JIT/inlining.

Reply:
Fair enough, makes sense.